packages/react/src/navigation-menu/trigger/NavigationMenuTrigger.tsx

```diff
@@ -129,76 +157,64 @@ export const NavigationMenuTrigger = React.forwardRef(function NavigationMenuTri
       sizeFrame2.request(() => {
         popupElement.style.setProperty(NavigationMenuPopupCssVars.popupWidth, `${nextWidth}px`);
         popupElement.style.setProperty(NavigationMenuPopupCssVars.popupHeight, `${nextHeight}px`);
+        runOnceAnimationsFinish(setAutoSizes, abortControllerRef.current.signal);
       });
     });
   });
 
-  const setAutoSizes = useEventCallback(() => {
-    if (!popupElement) {
-      return;
+  React.useEffect(() => {
+    if (!open) {
+      stickIfOpenTimeout.clear();
+      sizeFrame1.cancel();
+      sizeFrame2.cancel();
+      prevSizeRef.current = { width: 0, height: 0 };
     }
+  }, [stickIfOpenTimeout, open, sizeFrame1, sizeFrame2]);
 
-    popupElement.style.setProperty(NavigationMenuPopupCssVars.popupWidth, 'auto');
-    popupElement.style.setProperty(NavigationMenuPopupCssVars.popupHeight, 'auto');
-  });
-
-  const runOnceAnimationsFinish = useAnimationsFinished({ current: popupElement }, value);
-
-  useIsoLayoutEffect(() => {
-    if (!positionerElement || !popupElement || !open) {
+  React.useEffect(() => {
+    if (!popupElement || typeof ResizeObserver !== 'function') {
       return undefined;
     }
 
-    sizeFrame1.request(() => {
-      sizeFrame2.request(setAutoSizes);
+    const resizeObserver = new ResizeObserver(() => {
+      prevSizeRef.current = {
+        width: popupElement.offsetWidth,
+        height: popupElement.offsetHeight,
+      };
     });
+
+    resizeObserver.observe(popupElement);
+
     return () => {
-      sizeFrame1.cancel();
-      sizeFrame2.cancel();
+      resizeObserver.disconnect();
     };
-  }, [open, popupElement, positionerElement, sizeFrame1, sizeFrame2, setAutoSizes]);
+  }, [popupElement]);
 
-  useIsoLayoutEffect(() => {
-    if (!positionerElement || !popupElement || !value) {
+  React.useEffect(() => {
+    if (!popupElement || !isActiveItem || typeof MutationObserver !== 'function') {
       return undefined;
     }
 
-    const ac = new AbortController();
-    sizeFrame1.request(() => {
-      sizeFrame2.request(() => {
-        runOnceAnimationsFinish(setAutoSizes, ac.signal);
-      });
+    const mutationObserver = new MutationObserver(() => {
+      if (!popupElement) {
+        return;
+      }
+
+      abortControllerRef.current.abort();
+      abortControllerRef.current = new AbortController();
```

Comment:
This works, but it seems more intuitive to me to create a new `AbortController` when a new one is needed - when `runOnceAnimationsFinish` is called.